gemd/ingest/impl.py

```diff
@@ -0,0 +1,208 @@
+
+    """
+    if process_name is None:
+        if process_template is None:
+            process_name = name
+        else:
+            process_name = process_template.name
+
+    my_process_spec = ProcessSpec(
+        name=process_name,
+        template=process_template
+    )
+
+    my_process_run = ProcessRun(
+        name=process_name,
+        spec=my_process_spec
+    )
+
+    my_mat_spec = MaterialSpec(
+        name=name,
+        process=my_process_spec,
+        template=material_template
+    )
+
+    my_mat_run = MaterialRun(
+        name=name,
+        process=my_process_run,
+        spec=my_mat_spec
+    )
+    return my_mat_run
+
+
+def add_edge(input_material: MaterialRun,
+             output_material: MaterialRun,
+             *,
+             name: str = None):
+    """
+    Connect two material-process spec-run quadruples with ingredients.
+
+    Parameters
+    ----------
+    input_material: MaterialRun
+        The `material` for the returned IngredientRun
+    output_material: MaterialRun
+        The `process` for the returned IngredientRun will be
+        `output_material.process`
+    name: str
+        The ingredient name.  Defaults to `input_material.name`.
+
+    Returns
+    --------
+    MaterialRun
+        A MaterialRun with linked processes, specs and templates
+
+    """
+    if name is None:
+        name = input_material.name
+    my_ingredient_spec = IngredientSpec(name=name,
+                                        process=output_material.spec.process,
+                                        material=input_material.spec
+                                        )
+    my_ingredient_spec = IngredientRun(spec=my_ingredient_spec,
+                                       process=output_material.process,
+                                       material=input_material
+                                       )
+    return my_ingredient_spec
+
+
+def add_measurement(material: MaterialRun,
+                    name: str = None,
+                    *,
+                    template: MeasurementTemplate = None
+                    ):
+    """
+    Add a measurement run-spec set to a MaterialRun.
+
+    Parameters
+    ----------
+    material: MaterialRun
+        The `material` for the returned MeasurementRun
+    name: str
+        The name of the measurement.  Defaults to
+        `template.name` if `template` is defined.
+    template: MeasurementTemplate
+        The MeasurementTemplate.
+
+    Returns
+    --------
+    MeasurementRun
+        A MeasurementRun with linked material, spec and template
+
+    """
+    if name is None:
+        if template is None:
+            raise ValueError("Either a name or a template must be provided")
+        else:
+            name = template.name
+    my_measurement_spec = MeasurementSpec(name, template=template)
+    my_measurement_run = MeasurementRun(name, spec=my_measurement_spec, material=material)
+
+    return my_measurement_run
```

Comment:
How does the user assert the value of the measurement? I suppose it can be added later, but that seems like an important enough part of adding the measurement to do it inside this helper method.

Reply:
I would seem natural to let a user hand in a property at the same time as creating a measurement (having a measurement w/o any properties would be weird).  Since this would be an implicit `add_attribute`, you'd like to see the addition of a value and a property_template argument?

Reply:
> Since this would be an implicit add_attribute, you'd like to see the addition of a value and a property_template argument?

The user could instead provide an `Attribute` (or a list of `Attributes`) directly.

Providing a `value` and `property_template` means that it makes use of the existing logic in `add_attribute`. There's less code duplication and the user doesn't have to construct the `Attribute`, which can be nice if they want to use a primitive. The downside is that the user can only assign a single value to the measurement. Additional values would have to go through `add_attribute`, which is a clunky workflow. This would only be really valuable if most measurements have a single `Attribute`.

Providing a list of `Attributes` means that the measurement can be constructed in a single call, but the user has to make all of the `Attribute`s. There would also be some additional logic to assign each `Attribute` to the appropriate collection.

Doing nothing means that the workflow is more verbose but it's consistent. You always add measurements and then add attributes.

Reply:
I added a `make_attribute` method and an `attributes` argument to add_measurement to scratch this itch and try to still keep methods simple and consistent.  Let me know if the change makes sense.